File: design-patterns/go/facade/travel/car.go
// car.go

package main

import (
	"fmt"
	"math/rand"
)

type Car struct {
}

func NewCar() (car *Car) {
	car = &Car{}
	return
}

func (car *Car) GetDistanceTravelled() (float64) {
	return float64((rand.Intn((10000-100)*10 + 1) + 100*10) / 10.0)
}

func (car *Car) GoLeft() {
	fmt.Println("Go Left: ←")
}

func (car *Car) GoRight() {
	fmt.Println("Go Right: →")
}

func (car *Car) GoStraight() {
	fmt.Println("Go Straight: ↑")
}

func (car *Car) GoBack() {
	fmt.Println("Go Back: ↓")
}

func (car *Car) StartEngine() {
	fmt.Println("Start Engine")
}

func (car *Car) StopEngine() {
	fmt.Println("Stop Engine")
}

File: design-patterns/ts/facade/travel/car.ts
// car.ts

class Car {
    startEngine(): void {
        console.log("Start Engine");
    }

    stopEngine(): void {
        console.log("Stop Engine");
    }

    goStraight(): void {
        console.log("Go Straight: ↑");
    }

    goLeft(): void {
        console.log("Go Left: ←");
    }

    goRight(): void {
        console.log("Go Right: →");
    }

    getDistanceTravelled(): number {
        // This is some random calculation for demo purpose
        return ((Math.random() * ((10000 - 100) * 10 + 1)) + 100 * 10) / 10.0;
    }
}

export default Car;